File: src/services/auth/authService.ts
import type { AppError, Unsubscribe } from "../../shared/contracts";

export type AuthUser = {
  uid: string;
  displayName: string;
  email?: string;
  photoURL?: string;
  isAnonymous: boolean;
};

export type AuthStateListener = (user: AuthUser | null) => void;
export type AuthErrorListener = (error: AppError) => void;

export interface AuthService {
  subscribe(listener: AuthStateListener, onError: AuthErrorListener): Unsubscribe;
  ensureAnonymousSession(): Promise<AuthUser>;
  linkGoogleAccount(): Promise<AuthUser>;
}

File: frontend/lib/services/auth_service.dart
final class AuthUser {
  const AuthUser({
    required this.uid,
    required this.displayName,
    required this.isAnonymous,
    this.email,
    this.photoUrl,
  });

  final String uid;
  final String displayName;
  final String? email;
  final String? photoUrl;
  final bool isAnonymous;
}

abstract interface class AuthService {
  Stream<AuthUser?> authStateChanges();

  Future<AuthUser> ensureAnonymousSession();

  Future<AuthUser> linkGoogleAccount();
}
